text/0763-asset-importing-spec.md

```diff
@@ -0,0 +1,200 @@
+
+Today, the ecosystem mostly relies on [broccoli-asset-rev](https://github.com/ember-cli/broccoli-asset-rev). But broccoli-asset-rev has some major architectural problems and it doesn't take advantage of the newer capabilities we have in ember-auto-import and embroider.
+
+The biggest problem with broccoli-asset-rev is that it cannot achieve full correctness because it tries to detect inter-file references within arbitrary source code that is not statically analyzable. It's easy to accidentally refactor working code into code that works in development and test and then fails in production! For example, assuming broccoli-asset-rev is configured to fingerprint PNG assets, this refactor will look correct until you get into production and then have broken images:
+
+```diff
+class extends Component {
+-  iconURL = "/icons/happy.png"
++  get iconURL() {
++    return `/icons/${this.args.emotion}.png`
++  }
+}
+```
+
+The regular expressions that broccoli-asset-rev uses to do reference rewriting are effectively un-editable at this point. Any possible change you'd make to them breaks somebody's app.
+
+Another problem with broccoli-asset-rev is that it's "push based", meaning it handles files because they exist, rather than handling files because somebody is actually consuming them. It can't detect unused assets and it can't detect missing assets at build time. It exposes the underlying implementation of how assets get delivered, making it hard to change later without breaking app code.
+
+And because broccoli-asset-rev exposes assumptions about the exact layout of the final build output, it's not compatible with Embroider. Embroider allows different Javascript packagers to optimize the app delivery in any spec-compliant way. For example, small images might get inlined as `data:` URIs, CSS `@import()` might get inlined or not, etc. All these concerns really need to be integrated into the Javascript packager, because asset handling affects app-wide hashing & fingerprinting.
+
+In contrast, a pull-based design that lets code declare what assets it needs and then not worry about how those assets will get delivered is safer and easier to change in the future.
+
+## Detailed design
+
+Inter-file references take many forms. Here is a guide to inter-file references organized by which kind of file is doing the referring ("From Type") and which type of file it's referring to ("To Type"). Only the row containing **Proposed** is new in this RFC:
+
+From Type | To Type | Example Syntax | Semantics
+--------------------|-----------------------|--------|---
+JS | JS | `import { stuff } from 'thing';`| ECMA with NodeJS resolving rules
+JS | CSS | `import 'thing/style.css';` | [Embroider Spec CSS Rule](https://github.com/emberjs/rfcs/blob/master/text/0507-embroider-v2-package-format.md#css) (1)
```

Comment:
How would that play together with [native CSS modules](https://www.chromestatus.com/feature/5948572598009856) recently shipped as an experimental feature in Chrome?

I know there is much discussion around that spec. I don't have enough insights myself into bundling in general, CSS-in-JS etc. But I think we should consider how the decision made here may affect a migration path.

Reply:
Thanks for pointing that out, I had not been following that proposal. It relies on [import assertions](https://github.com/tc39/proposal-import-assertions) syntax, which is currently in stage3 at tc39.

Because that proposal relies on new syntax, it wouldn't conflict with anything in this proposal as written. We could adopt that proposal when it stabilizes. But it's worth asking if we shouldn't just jump directly to that proposal. Mostly that comes down to practical questions of how likely it is to stabilize without further changes and whether enough existing tooling supports it.

For example, webpack seems to be able to parse the new syntax, but doesn't let you write rules dependent on the import assertions, as far as I can tell.  TypeScript doesn't have an implementation yet. Unsure about eslint.

Reply:
> Because that proposal relies on new syntax, it wouldn't conflict with anything in this proposal as written.

Thanks for clarifying.

> But it's worth asking if we shouldn't just jump directly to that proposal.

I'm not sure if we can at this point in time. If I got the proposal right, importing CSS does not have side-effects. Instead the default export is a [constructable style sheet](https://developer.mozilla.org/en-US/docs/Web/API/CSSStyleSheet), which must be applied explicitly to either `document` or a shadow root. A style sheet can be applied using `adoptedStyleSheets` API. But that one is not supported by Firefox and Safari yet [accordingly to caniuse](https://caniuse.com/mdn-api_document_adoptedstylesheets).

Reply:
Whether browsers implement is not really the limiting factor, because we can transpile / polyfill as needed.

The limiting factor is whether developer tooling works well enough on your code as-authored that you'd want to use these features in it.